### PR TITLE
Add sig api with picnic

### PR DESCRIPTION
**Pull request by christianpaquin**
This pull request proposes a new signature API modeled after the kex api, and adds the [Picnic](https://github.com/Microsoft/Picnic) signature scheme.

Picnic must currently be downloaded and set up (to generate parameters) using a script (download-and-setup-picnic.sh). We can migrate to a git submodule at a later time, after this PR is reviewed. 

The Visual Studio build has not yet been modified to support Picnic; this will be done after this PR is reviewed.

### CODE REVIEW

**Comment by smashra** [comment]
I'm getting following error, please fix:
external/Picnic-master/LowMC.c: In function 'getAllRandomness':
external/Picnic-master/LowMC.c:209:20: error: storage size of 'ctx' isn't known
     EVP_CIPHER_CTX ctx;
                    ^~~
external/Picnic-master/LowMC.c:209:20: error: unused variable 'ctx' [-Werror=unused-variable]
external/Picnic-master/LowMC.c: In function 'cleanup_EVP':
external/Picnic-master/LowMC.c:245:5: error: 'ERR_remove_state' is deprecated [-Werror=deprecated-declarations]
     ERR_remove_state(0);
     ^~~~~~~~~~~~~~~~
In file included from /usr/local/include/openssl/bn.h:31:0,
                 from /usr/local/include/openssl/asn1.h:24,
                 from /usr/local/include/openssl/objects.h:15,
                 from /usr/local/include/openssl/evp.h:28,
                 from external/Picnic-master/LowMC.c:32:
/usr/local/include/openssl/err.h:249:1: note: declared here
 DEPRECATEDIN_1_0_0(void ERR_remove_state(unsigned long pid))
 ^
cc1: all warnings being treated as errors

**Comment by tlepoint** [comment]
Apparently, picnic is in master?

**Comment by dstebila** [comment]
Yeah there's been a bit of confusion about this PR and what should be added.  I'm working with @smashra to try to sort it out and will update you on here in the next few days.

**Comment by smashra** [comment]
Removing make and build steps from download-and-setup-picnic.sh as they are being taken care of from the main Makefile with configure options in configure.ac.

**Comment by smashra** [comment]
Made 
#include <graycode.h>
to
#include <m4ri/graycode.h>
in preprocessmatrices.c

for making --with-m4ri-dir option consistent with others.

Please update your branch.

**Comment by smashra** [comment]
Can you please create another pull request to fix this:
====================
Made
#include <graycode.h>
to
#include <m4ri/graycode.h>
in preprocessmatrices.c

for making --with-m4ri-dir option consistent with others.
======================

Such that I can check in picnic code.

**Comment by smashra** [comment]
Sorry, the bold and caps was a mistake, totally unintentional.

**Comment by smashra** [comment]
Please fix these warning:


src/sig_picnic/external/Picnic-master/preprocessMatrices.c:70:23: error: comparison between signed and unsigned integer expressions [-Werror=sign-compare]
     for (int i = 0; i < numRounds; i++) {
                       ^
src/sig_picnic/external/Picnic-master/preprocessMatrices.c:72:27: error: comparison between signed and unsigned integer expressions [-Werror=sign-compare]
         for (int j = 0; j < stateBits / t; j++) {
                           ^
src/sig_picnic/external/Picnic-master/preprocessMatrices.c:80:23: error: comparison between signed and unsigned integer expressions [-Werror=sign-compare]
     for (int i = 0; i < numRounds; i++) {
                       ^
src/sig_picnic/external/Picnic-master/preprocessMatrices.c:82:27: error: comparison between signed and unsigned integer expressions [-Werror=sign-compare]
         for (int j = 0; j < stateBits / t; j++) {
                           ^
src/sig_picnic/external/Picnic-master/preprocessMatrices.c:93:23: error: comparison between signed and unsigned integer expressions [-Werror=sign-compare]
     for (int i = 0; i < numRounds; i++) {
                       ^
src/sig_picnic/external/Picnic-master/preprocessMatrices.c:94:27: error: comparison between signed and unsigned integer expressions [-Werror=sign-compare]
         for (int j = 0; j < stateBits / t; j++) {
                           ^
src/sig_picnic/external/Picnic-master/preprocessMatrices.c:105:19: error: comparison between signed and unsigned integer expressions [-Werror=sign-compare]
     for (i = 0; i < numRounds; i++) {
                   ^
src/sig_picnic/external/Picnic-master/preprocessMatrices.c:107:23: error: comparison between signed and unsigned integer expressions [-Werror=sign-compare]
         for (j = 0; j < stateBits / t; j++) {
                       ^
src/sig_picnic/external/Picnic-master/preprocessMatrices.c:111:31: error: comparison between signed and unsigned integer expressions [-Werror=sign-compare]
                 for (l = 0; l < stateWords; l++) {
                               ^
src/sig_picnic/external/Picnic-master/preprocessMatrices.c:113:27: error: comparison between signed and unsigned integer expressions [-Werror=sign-compare]
                     if (l != stateWords - 1) {
                           ^~
src/sig_picnic/external/Picnic-master/preprocessMatrices.c:126:19: error: comparison between signed and unsigned integer expressions [-Werror=sign-compare]
             if (j != (stateBits / t) - 1) {
                   ^~
src/sig_picnic/external/Picnic-master/preprocessMatrices.c:132:15: error: comparison between signed and unsigned integer expressions [-Werror=sign-compare]
         if (i != numRounds - 1) {
               ^~
src/sig_picnic/external/Picnic-master/preprocessMatrices.c:138:23: error: comparison between signed and unsigned integer expressions [-Werror=sign-compare]
     for (int i = 0; i < numRounds; i++) {
                       ^
src/sig_picnic/external/Picnic-master/preprocessMatrices.c:139:27: error: comparison between signed and unsigned integer expressions [-Werror=sign-compare]
         for (int j = 0; j < stateBits / t; j++) {
                           ^
src/sig_picnic/external/Picnic-master/preprocessMatrices.c: In function ‘writeRoundConstants’:
src/sig_picnic/external/Picnic-master/preprocessMatrices.c:152:23: error: comparison between signed and unsigned integer expressions [-Werror=sign-compare]
     for (int i = 0; i < numRounds; i++) {
                       ^
src/sig_picnic/external/Picnic-master/preprocessMatrices.c: In function ‘generatelookupTables’:
src/sig_picnic/external/Picnic-master/preprocessMatrices.c:180:5: error: enumeration value ‘PARAMETER_SET_INVALID’ not handled in switch [-Werror=switch]
     switch (picnic_params) {
     ^~~~~~
src/sig_picnic/external/Picnic-master/preprocessMatrices.c:180:5: error: enumeration value ‘PARAMETER_SET_MAX_INDEX’ not handled in switch [-Werror=switch]

**Comment by dstebila** [comment]
Since Picnic-master is an external dependency, can we compile that directory without -Werror while the upstream provider considers whether to adopt the changes or not?

**Comment by smashra** [comment]
Ok.

**Comment by smashra** [comment]
I'm already getting name collision error:

./.libs/liboqs.a(libpicnic_la-LowMC.o): In function `verify':
/home/shravan/workspace/delete_picnic/liboqs/src/sig_picnic/external/Picnic-master/LowMC.c:1050: multiple definition of `verify'
./.libs/liboqs.a(libkyber_la-kex_mlwe_kyber.o):/home/shravan/workspace/delete_picnic/liboqs/src/kex_mlwe_kyber/verify.c:10: first defined here
collect2: error: ld returned 1 exit status
make[2]: *** [Makefile:732: test_sig] Error 1

Douglas, we discussed to not enable picnic by default but why does 
src/kex_mlwe_kyber/verify.c
has a global function verify() and still it has been accepted in master and enabled by default.

**Comment by dstebila** [comment]
@smashra Apparently there were non-namespaced names added in the merge of Kyber (#131) and these weren't caught by Travis since Travis isn't correctly catching all failures (open issue #130).  So I think the sequence of operations would be:

1. Prepare a pull request that fixes #130, which will mean current builds fail.
2. Prepare a pull request that builds on the pull request in 1. and that fixes the non-namespaced symbols in Kyber.
3. Merge both of those to master.
4. Update this pull request.

**Comment by christianpaquin** [comment]
@smashra, the last commit #6153022 removed the Picnic matrices preprocessing in the download-and-setup-picnic.sh script. How do you propose we setup Picnic after downloading it?

**Comment by smashra** [comment]
It's done through root or main make file.am.

You'll see when you run make test.

**Comment by christianpaquin** [comment]
OK, works for me. Is there a reason why the matrices preprocessing is not part of the make (vs. make test)? I updated the README to include "make test" as a step when building Picnic. When are you planning to merge this with master? I'm working on fixing the Windows build; if this is going in soon, I'll wait to also include the sig api and Picnic.

**Comment by smashra** [comment]
The reason is consistency on my end with respect to passing compiler and
directory flags, considering that it is an external component, such that it
appears more like an internal project from build perspective.

Please wait for an hour, I'm in the process of merging some mac related
updates.

Thanks for waiting.

**Comment by smashra** [comment]
@Christian: I have merged my changes, please go ahead.

**Comment by christianpaquin** [comment]
I tested the build with your commits @smashra; looks good.

**Comment by smashra** [comment]
Great!

**Comment by tlepoint** [comment]
I get a warning during compilation on macOS

```
  CC       src/sig_picnic/external/Picnic-master/pp_matrices-preprocessMatrices.o
src/sig_picnic/external/Picnic-master/preprocessMatrices.c:180:13: warning: enumeration values
      'PARAMETER_SET_INVALID' and 'PARAMETER_SET_MAX_INDEX' not handled in switch [-Wswitch]
    switch (picnic_params) {
            ^
1 warning generated.
```

**Comment by dstebila** [comment]
@smashra There's still a failure in one of the macOS builds due to a header not being in the right location.  Can you look into that?

**Comment by smashra** [comment]
Is it beacause of the merge? I'll look into it.

**Comment by smashra** [comment]
I'm looking at

https://travis-ci.org/open-quantum-safe/liboqs/builds/254945234?utm_source=github_status&utm_medium=notification

where one of the mac has sig_picnic enabled.

I don't see any errors.

Where should I look ?

Sincerely
Shravan

File: configure.ac [hunk 1]
```diff
@@ -96,10 +96,16 @@ ARG_ENABL_SET([kex-ntru], [enable KEX-NTRU.])
...
+ARG_ENABL_SET([picnic], [enable PICNIC.])
+AM_CONDITIONAL([picnic], [test "x$picnic" = xtrue])
+AM_CONDITIONAL([USE_PICNIC], [test "x$picnic" = xtrue])
+
 ARG_ENABL_SET([kex-sidh-iqc-ref], [enable KEX-SIDH-IQC-REF.])
 AM_CONDITIONAL([kex_sidh_iqc_ref], [test "x$kex_sidh_iqc_ref" = xtrue])
 AM_CONDITIONAL([USE_SIDH_IQC], [test "x$kex_sidh_iqc_ref" = xtrue])
 
+
+AM_CPPFLAGS="-g -std=gnu11 -Wno-unused-function -Werror -Wpedantic -Wall -Wextra -DCONSTANT_TIME"
```
Comment (dstebila) [inline configure.ac]: Why is this line changing?

File: configure.ac [hunk 2]
```diff
@@ -120,6 +126,35 @@ case $host_os in
         ;;
 esac
 
+AC_ARG_WITH(
```
Comment (dstebila) [inline configure.ac]: Why are lines 129-149 changing? Doesn't seem related to the Picnic commit.

File: configure.ac [hunk 3]
```diff
@@ -129,6 +164,8 @@ SRCDIR=${SRCDIR}" src/crypto/aes src/crypto/rand src/crypto/sha3 src/crypto/rand
 
 # KEX
 SRCDIR=${SRCDIR}" src/kex"
+SRCDIR=" src/common src/crypto/aes src/kex src/sig src/crypto/rand src/crypto/sha3"
+SRCDIR=${SRCDIR}" src/crypto/rand_urandom_aesctr src/crypto/rand_urandom_chacha20"
```
Comment (dstebila) [inline configure.ac]: Why is this line changing? Doesn't seem related to the Picnic commit.

File: configure.ac [hunk 4]
```diff
@@ -137,6 +174,11 @@ SRCDIR=${SRCDIR}" src/kex_rlwe_msrln16"
 SRCDIR=${SRCDIR}" src/kex_rlwe_newhope"
 SRCDIR=${SRCDIR}" src/kex_sidh_cln16"
 
+if test x"$kex_ntru" = x"true"; then
+  AM_CPPFLAGS=${AM_CPPFLAGS}" -DENABLE_NTRU"
+  SRCDIR=${SRCDIR}" src/kex_ntru"
+fi
+
```
Comment (dstebila) [inline configure.ac]: Why are these lines changing? Doesn't seem related to the Picnic commit.

File: configure.ac [hunk 5]
```diff
@@ -158,6 +200,12 @@ if test x"$kex_code_mcbits" = x"true"; then
 fi
 if test x"$kex_sidh_iqc_ref" = x"true"; then
   AM_CPPFLAGS=${AM_CPPFLAGS}" -DENABLE_SIDH_IQC_REF"
+  SRCDIR=${SRCDIR}" src/kex_sidh_iqc_ref"
```
Comment (dstebila) [inline configure.ac]: Why is this line changing? Doesn't seem related to the Picnic commit.

File: set-picnic-var.sh [hunk 6]
```diff
@@ -0,0 +1,4 @@
+#!/bin/sh
+export PICNIC_PARAMS_PATH=$PWD/src/sig_picnic/external/Picnic-master/precomputed_data/
+echo $PICNIC_PARAMS_PATH
+
```
Comment (dstebila) [inline set-picnic-var.sh]: What's the purpose of this script? There's no explanation in the README of its role.

File: src/crypto/rand/rand.h [hunk 7]
```diff
@@ -20,7 +20,7 @@ typedef struct OQS_RAND OQS_RAND;
 /**
  * OQS PRNG object
  */
-typedef struct OQS_RAND {
+struct OQS_RAND {
```
Comment (dstebila) [inline src/crypto/rand/rand.h]: Why is this changing?

File: configure.ac (hunk 1 above)
Comment (smashra) [inline configure.ac]: It's a duplicate, there is another AM_CPPFLAGS below. Needs to be removed. Merge artifact.

File: set-picnic-var.sh (hunk 6 above)
Comment (dstebila) [inline set-picnic-var.sh]: Okay, so can we delete it?

Comment (smashra) [inline set-picnic-var.sh]: Yes, we can delete it.

File: src/crypto/rand/rand.h (hunk 7 above)
Comment (dstebila) [inline src/crypto/rand/rand.h]: This should be
~~~
typedef struct OQS_RAND OQS_RAND;
struct OQS_RAND { ... };
~~~

Comment (smashra) [inline src/crypto/rand/rand.h]: The above change already exists in rand.h.

Comment (dstebila) [inline src/crypto/rand/rand.h]: Ah, I see, thanks.
